### PR TITLE
Docs: Getting Started Dependencies

### PR DESCRIPTION
**Pull request by ax3l**
Finally document what one needs to install to use Spack on Linux and Mac :-)

With <3 for minimal container users (e.g., issues like #32479) and my colleagues with their fancy Macs.
Screenshot from 2022-09-01 12-03-42

### CODE REVIEW

File: lib/spack/docs/spack.yaml [hunk 1]
```diff
@@ -18,6 +18,8 @@ spack:
   - "py-sphinx@3.4:4.1.1,4.1.3:"
   - py-sphinxcontrib-programoutput
   - py-docutils@:0.16
+  # TODO: needs to be packaged
+  - py-sphinx-design
```
Comment (ax3l) [inline lib/spack/docs/spack.yaml]: Via #32482
```suggestion
  - py-sphinx-design
```

File: lib/spack/docs/getting_started.rst [hunk 2]
```diff
@@ -23,8 +23,31 @@ be present on the machine where Spack is run:
...
+A build matrix showing which packages are working on which systems is shown below.
+
+.. tab-set::
+
+   .. tab-item:: Debian/Ubuntu
+
+      .. code-block:: console
+
+         apt update
+         apt install build-essential ca-certificates coreutils curl environment-modules gcc gfortran g++ git gpg lsb-release python3 unzip zip
```
Comment (cgmb) [inline lib/spack/docs/getting_started.rst]: `gcc` and `g++` are part of `build-essential`, so they don't need to be listed separately. I also always install `python3-distutils`. I can't remember why, but I recall it being necessary on occasion...

Comment (adamjstewart) [reply lib/spack/docs/getting_started.rst]: Spack used to use distutils but no longer does. Some packages depends on distutils, but I believe pip patches those out so distutils doesn't actually get used. @pradyunsg?

Comment (cgmb) [reply lib/spack/docs/getting_started.rst]: Ah. I know why. `spack style` will fail without `python3-distutils python3-dev`.

Comment (adamjstewart) [reply lib/spack/docs/getting_started.rst]: What's the error message?

Comment (cgmb) [reply lib/spack/docs/getting_started.rst]: Without python3-distutils:
```
==> Installing py-pip-22.1.2-rltemoksgf7tbslmak4yztuqseppc724
==> No binary for py-pip-22.1.2-rltemoksgf7tbslmak4yztuqseppc724 found: installing from source
==> Using cached archive: /root/spack/var/spack/cache/_source-cache/archive/a3/a3edacb89022ef5258bf61852728bf866632a394da837ca49eb4303635835f17
==> No patches needed for py-pip
==> py-pip: Executing phase: 'install'
==> Error: ProcessError: Command exited with status 1:
    '/usr/bin/python3.8' '/tmp/root/spack-stage/spack-stage-py-pip-22.1.2-rltemoksgf7tbslmak4yztuqseppc724/spack-src/pip-22.1.2-py3-none-any.whl/pip' '-vvv' '--no-input' '--no-cache-dir' '--disable-pip-version-check' 'install' '--no-deps' '--ignore-installed' '--no-build-isolation' '--no-warn-script-location' '--no-index' '--prefix=/root/.spack/bootstrap/store/linux-ubuntu20.04-x86_64/gcc-9.4.0/py-pip-22.1.2-rltemoksgf7tbslmak4yztuqseppc724' '/tmp/root/spack-stage/spack-stage-py-pip-22.1.2-rltemoksgf7tbslmak4yztuqseppc724/spack-src/pip-22.1.2-py3-none-any.whl'
See build log for details:
  /tmp/root/spack-stage/spack-stage-py-pip-22.1.2-rltemoksgf7tbslmak4yztuqseppc724/spack-build-out.txt
```

Without python3-dev:
```
==> Error: cannot bootstrap any of the isort executables from spec "py-isort@4.3.5: %gcc target=x86_64" due to the following failures:
github-actions-v0.3 raised FileNotFoundError: [Errno 2] No such file or directory: '/root/spack/share/spack/bootstrap/github-actions-v0.3/py-isort.json'
github-actions-v0.2 raised ValueError: source is not trusted
github-actions-v0.1 raised ValueError: source is not trusted
spack-install raised NoHeadersError: Unable to locate python headers in any of these locations:
    /usr/include/python3.8
/usr/include/3.8
/usr/Headers
Run `spack --debug ...` for more detailed errors
```

Comment (adamjstewart) [reply lib/spack/docs/getting_started.rst]: "without python3-dev" makes sense, can you share the build log for "without python3-distutils"?

Comment (cgmb) [reply lib/spack/docs/getting_started.rst]: On closer look, I notice that `python3-dev` depends on `python3-distutils` so I suppose specifying both is redundant. Still, there definitely is a step that requires `python3-distutils` itself, as installing that package allows the build to advance a little further than when neither is installed. This is the log when running `spack style` without either package on Ubuntu 20.04:
```
# cat /tmp/root/spack-stage/spack-stage-py-pip-22.1.2-rltemoksgf7tbslmak4yztuqseppc724/spack-build-out.txt
==> py-pip: Executing phase: 'install'
==> [2022-09-09-02:51:57.531750] '/usr/bin/python3.8' '/tmp/root/spack-stage/spack-stage-py-pip-22.1.2-rltemoksgf7tbslmak4yztuqseppc724/spack-src/pip-22.1.2-py3-none-any.whl/pip' '-vvv' '--no-input' '--no-cache-dir' '--disable-pip-version-check' 'install' '--no-deps' '--ignore-installed' '--no-build-isolation' '--no-warn-script-location' '--no-index' '--prefix=/root/.spack/bootstrap/store/linux-ubuntu20.04-x86_64/gcc-9.4.0/py-pip-22.1.2-rltemoksgf7tbslmak4yztuqseppc724' '/tmp/root/spack-stage/spack-stage-py-pip-22.1.2-rltemoksgf7tbslmak4yztuqseppc724/spack-src/pip-22.1.2-py3-none-any.whl'
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/tmp/root/spack-stage/spack-stage-py-pip-22.1.2-rltemoksgf7tbslmak4yztuqseppc724/spack-src/pip-22.1.2-py3-none-any.whl/pip/__main__.py", line 29, in <module>
  File "<frozen zipimport>", line 259, in load_module
  File "/tmp/root/spack-stage/spack-stage-py-pip-22.1.2-rltemoksgf7tbslmak4yztuqseppc724/spack-src/pip-22.1.2-py3-none-any.whl/pip/_internal/cli/main.py", line 9, in <module>
  File "<frozen zipimport>", line 259, in load_module
  File "/tmp/root/spack-stage/spack-stage-py-pip-22.1.2-rltemoksgf7tbslmak4yztuqseppc724/spack-src/pip-22.1.2-py3-none-any.whl/pip/_internal/cli/autocompletion.py", line 10, in <module>
  File "<frozen zipimport>", line 259, in load_module
  File "/tmp/root/spack-stage/spack-stage-py-pip-22.1.2-rltemoksgf7tbslmak4yztuqseppc724/spack-src/pip-22.1.2-py3-none-any.whl/pip/_internal/cli/main_parser.py", line 8, in <module>
  File "<frozen zipimport>", line 259, in load_module
  File "/tmp/root/spack-stage/spack-stage-py-pip-22.1.2-rltemoksgf7tbslmak4yztuqseppc724/spack-src/pip-22.1.2-py3-none-any.whl/pip/_internal/cli/cmdoptions.py", line 24, in <module>
  File "<frozen zipimport>", line 259, in load_module
  File "/tmp/root/spack-stage/spack-stage-py-pip-22.1.2-rltemoksgf7tbslmak4yztuqseppc724/spack-src/pip-22.1.2-py3-none-any.whl/pip/_internal/cli/parser.py", line 12, in <module>
  File "<frozen zipimport>", line 259, in load_module
  File "/tmp/root/spack-stage/spack-stage-py-pip-22.1.2-rltemoksgf7tbslmak4yztuqseppc724/spack-src/pip-22.1.2-py3-none-any.whl/pip/_internal/configuration.py", line 26, in <module>
  File "<frozen zipimport>", line 259, in load_module
  File "/tmp/root/spack-stage/spack-stage-py-pip-22.1.2-rltemoksgf7tbslmak4yztuqseppc724/spack-src/pip-22.1.2-py3-none-any.whl/pip/_internal/utils/logging.py", line 29, in <module>
  File "<frozen zipimport>", line 259, in load_module
  File "/tmp/root/spack-stage/spack-stage-py-pip-22.1.2-rltemoksgf7tbslmak4yztuqseppc724/spack-src/pip-22.1.2-py3-none-any.whl/pip/_internal/utils/misc.py", line 42, in <module>
  File "<frozen zipimport>", line 259, in load_module
  File "/tmp/root/spack-stage/spack-stage-py-pip-22.1.2-rltemoksgf7tbslmak4yztuqseppc724/spack-src/pip-22.1.2-py3-none-any.whl/pip/_internal/locations/__init__.py", line 14, in <module>
  File "<frozen zipimport>", line 259, in load_module
  File "/tmp/root/spack-stage/spack-stage-py-pip-22.1.2-rltemoksgf7tbslmak4yztuqseppc724/spack-src/pip-22.1.2-py3-none-any.whl/pip/_internal/locations/_distutils.py", line 9, in <module>
ModuleNotFoundError: No module named 'distutils.cmd'
```

Comment (pradyunsg) [reply lib/spack/docs/getting_started.rst]: On Debian and friends, you should include python3-venv and python3-distutils in the set of packages to be installed — Debian breaks up the Python standard library for some reason that I don’t fully understand and that breaks various packages and pip’s install mechanisms prior to Python 3.10, even though pip uses setuptools’ injection of distutils to ensure that packages with import distutils do the right thing in their build.

IIRC, there’s also a python3-full on newer versions of Debian but I’m not sure about the availability story if that in Debian/Ubuntu (i.e. versions).

**Review by alalazo — CHANGES_REQUESTED** [review]
Can we remove the CSV table at the beginning of the section in favor of this matrix?

File: lib/spack/docs/getting_started.rst (hunk 2 above)
Comment (ax3l) [reply lib/spack/docs/getting_started.rst]: Thank you all!

What's the result on `python3-dev` - do we need that too or is `python3-distutils` sufficient?

**Comment by ax3l** [comment]
I thought the CSV table is a platform agnostic dependency overview... Thus I kept it for now.

File: lib/spack/docs/getting_started.rst [hunk 3]
```diff
@@ -23,8 +23,31 @@ be present on the machine where Spack is run:
...
+      .. code-block:: console
+
+         apt update
+         apt install build-essential ca-certificates coreutils curl environment-modules gfortran git gpg lsb-release python3 python3-distutils python3-venv unzip zip
+
+   .. tab-item:: RHEL
+
+      .. note::
+
+         TODO
```
Comment (ax3l) [inline lib/spack/docs/getting_started.rst]: According to the docker images here: https://github.com/spack/spack/actions/runs/3216648825
```suggestion
         yum update -y
         yum install -y epel-release
         yum update -y
         yum --enablerepo epel groupinstall -y "Development Tools"
         yum --enablerepo epel install -y curl findutils gcc-c++ gcc gcc-gfortran git gnupg2 hostname iproute make patch python3 python3-pip python3-setuptools unzip
         python3 -m pip install boto3
```

File: lib/spack/docs/getting_started.rst (hunk 2 above)
Comment (ax3l) [reply lib/spack/docs/getting_started.rst]: Debian/Ubuntu: Should we also add for S3 support the
```bash
... python3-pip python3-setuptools
python3 -m pip install boto3
```

as in our docker images? https://github.com/spack/spack/actions/runs/3216648825

**Review by alalazo — APPROVED** [review]

File: lib/spack/docs/getting_started.rst (hunk 2 above)
Comment (pradyunsg) [reply lib/spack/docs/getting_started.rst]: -dev provides headers, which you need for compiling c extensions.

Comment (ax3l) [reply lib/spack/docs/getting_started.rst]: Yep, I don't think we need that in bootstrapping Spack... When a Python package is built in Spack, we first build a full Python itself again.

**Pull request merged**